routers/web/user/setting/oauth2_common.go

```diff
@@ -31,6 +31,7 @@ func (oa *OAuth2CommonHandlers) renderEditPage(ctx *context.Context) {
 func (oa *OAuth2CommonHandlers) AddApp(ctx *context.Context) {
 	form := web.GetForm(ctx).(*forms.EditOAuth2ApplicationForm)
 	if ctx.HasError() {
+		ctx.Flash.Error(ctx.GetErrMsg())
```

Comment:
Not sure whether the new code is right. But the old `HasError` has some strange behaviors:

![image](https://github.com/go-gitea/gitea/assets/2114189/ed9c20e0-c5ec-47e9-ae4c-4063fa2c0a9f)

Reply:
So I should add `ctx.Flash.Error(ctx.GetErrMsg())` to this function?

Reply:
No idea, adding more code like `ctx.Flash.Error(ctx.GetErrMsg())` doesn't seem right either. I haven't fully understand this "Flash message" problem yet.

The question in my mind is: why the old code `Flash.ErrorMsg = ...` doesn't work? What's the difference between it and the `ctx.Flash.Error(ctx.GetErrMsg())`

Reply:
I think I understand the problem now. Thank you very much.

The problem is: sometimes the Flash message is rendered in current request, sometimes the Flash message should be stored into cookie to be rendered in next request.

Your example's case is the first one, this PR's case is the second one.

The Flash message system is quite buggy ....